### DISCUSSION TITLE
Vaadin I18N Provider is not instantiated even though it should

### DISCUSSION THREAD

**Discussion by ckosmowski** [quarkusio/quarkus#38726] [Q&A] [answered]
I am trying to use a custom vaadin i18nProvider however it will never be picked up as a been:

```java
@VaadinServiceEnabled
@VaadinServiceScoped
public class CustomI18NProvider implements I18NProvider {

    @Override
    public List<Locale> getProvidedLocales() {
        return List.of(Locale.ENGLISH, Locale.GERMAN);
    }

    @Override
    public String getTranslation(String s, Locale locale, Object... objects) {
        return "Bla";
    }
}
```

During debugging i even get a ClassNotFoundException even though the class is there on the classpath. A lot of other beans load correctly, i tried a ton of combinations for annotations. 

I debugged into the QuarkusInstantiator and all requirements are met except that the bean is just not there. When i try to search for in in debugging i get the ClassNotFound described above.

**Comment by sberyozkin** [comment]
CC @phillip-kruger

**Comment by gsmet** [comment]
Are you using a Vaadin Quarkus extension? Is it with the JVM or as a native executable?

**Reply by ckosmowski** [reply]
I just checked: Yes the vaadin quarkus extension is used in the maven pom.

**Reply by ckosmowski** [reply]
I don't have enough understanding of the project to say for sure but it seems to me like the problem is rather that the class itsself is not picked up by the quarkus classloader than the identification and pickup as a vaadin compatible CDI bean.

**Reply by gsmet** [reply]
I was asking about native because when building a native executable, classes can be stripped from the executable. That's usually the role of the extension to make sure useful classes are kept around.
But that's not the case here because you are using the JVM.

**Comment by ckosmowski** [comment]
I got it working, or at least I have a workaround:

1. I needed an injection point for the I18NProvider even though I don't need to access it I needed to inject it anywhere, I think that makes the Quarkus class loader think that class might be important, and it loads it?! (correct me if I'm wrong).
1. When I did this, I got alerted by quarkus that no matching beans were found because qualifiers don't match. Seems the above annotations did not include the @Default qualifier so I had to add it. The class header now looks like this:

```java
@VaadinServiceEnabled
@VaadinServiceScoped
@Default
public class CustomI18NProvider implements I18NProvider {

```

**Reply by gsmet** [reply, marked as answer]
Yeah so your problem is that beans that look useless to our CDI layer are removed. That's why adding an injection point somewhere makes it work as when the injection point is detected, it says our CDI layer that the bean is consumed and shouldn't be removed.
Most probably, Vaadin is checking the beans dynamically without an injection point.

What you can do for the time being is add the `@Unremovable` annotation to your bean and drop the injection point you added.

I think it's also worth reporting this issue to the Vaadin extension author so that they make use of the `UnremovableBeanBuildItem` to mark the beans they actively consume as unremovable automatically.

HTH.

**Reply by ckosmowski** [reply]
Yes, that helps. With your hint, I also found the section about bean removing in the cdi-reference docs.

**Comment by gvirtuoso** [comment]
Still nowadays (Oct 27 of 2024) on Vaadin v 24.5.0 + Quakus 3.15.1 I had to add the `@Unremovable` annotation from `io.quarkus.arc.Unremovable` in order to clear the message.

```2024-10-27 23:02:07,849 INFO  [com.vaa.qua.QuarkusInstantiator] (Quarkus Main Thread) Can't find any @VaadinServiceScoped bean implementing 'I18NProvider'. Cannot use CDI beans for I18N, falling back to the default behavior.```

Implementation:
```
@VaadinServiceEnabled
@VaadinServiceScoped
@Unremovable
public class MessageProvider implements I18NProvider {
   ....
}
```

**Reply by gvirtuoso** [reply]
Anyways, besides the implementation above has solved the Log messages about a I18N implementation, I can't inject the Provider inside a View. (I'm using Quakus).  
That way my `MessageProvider` is always null.  
I tried to remove the Vaadin annotations and just set as an `@ApplicationScoped` but didn't work too.  

Eg:
```
package app.views.dashboard;

import app.enums.Language;
import app.providers.MessageProvider;
import com.vaadin.flow.component.html.H2;
import com.vaadin.flow.component.html.Image;
import com.vaadin.flow.component.html.Paragraph;
import com.vaadin.flow.component.html.Span;
import com.vaadin.flow.component.orderedlayout.VerticalLayout;
import com.vaadin.flow.router.Menu;
import com.vaadin.flow.router.Route;
import com.vaadin.flow.theme.lumo.LumoUtility.Margin;
import jakarta.inject.Inject;

@Route("")
@Menu(order = 0, icon = "line-awesome/svg/home-solid.svg")
public class DashboardView extends VerticalLayout {

    @Inject
    public MessageProvider messageProvider;

    public DashboardView() {

        setSpacing(false);

        String greeting = messageProvider.getTranslation("greeting", Language.getDefaultLocale());
        Span greetingSpan = new Span(greeting);
        add(greetingSpan);

        Image img = new Image("images/empty-plant.png", "placeholder plant");
        img.setWidth("200px");
        add(img);

        H2 header = new H2("This place intentionally left empty");
        header.addClassNames(Margin.Top.XLARGE, Margin.Bottom.MEDIUM);
        add(header);
        add(new Paragraph("It’s a place where you can grow your own UI 🤗"));

        setSizeFull();
        setJustifyContentMode(JustifyContentMode.CENTER);
        setDefaultHorizontalComponentAlignment(Alignment.CENTER);
        getStyle().set("text-align", "center");

    }

}

```

**Reply by gvirtuoso** [reply]
The only way it worked for me was like this:

- Annotate the MessageProvider class as @jakarta.enterprise.context.ApplicationScoped
    - (At this moment I had to ignore the i18n message on the startup log... who cares)
- On the DashboardView
    - Annotate with @jakarta.inject.Inject an attribute for the messageProvider
    - And even with the notation (which is not sufficient but required) I had to use the ARC static class to load a MessageProvider instance from Quarkus.
```
@Route("")
@Menu(order = 0, icon = "line-awesome/svg/home-solid.svg")
public class DashboardView extends VerticalLayout {

    @Inject
    public MessageProvider messageProvider;

    public DashboardView() {

        MessageProvider messageProvider = Arc.container().instance(MessageProvider.class).get();

        String greeting = messageProvider.getTranslation("greeting", Language.getDefaultLocale());
        Span greetingSpan = new Span(greeting);
        add(greetingSpan);

    }
    
```

With all the steps above, the injection of my i18n system worked properly. (Not sure if it would be the right way, but worked)
